### PR TITLE
add tdnn xent and tdnn+chain recipe for librispeech

### PR DESCRIPTION
**Pull request by freewym**
Create this PR to the master branch, to replace #645

### CODE REVIEW

**Comment by vijayaditya** [comment]
Why is this new PR needed what changed from #645 ?

**Comment by freewym** [comment]
#645 is for chain branch

File: egs/librispeech/s5/local/chain/run_tdnn_6z_discriminative.sh [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+set -o pipefail
+set -e
+# this is run_discriminative.sh
+
+# This script does discriminative training on top of chain nnet3 system.
+# note: this relies on having a cluster that has plenty of CPUs as well as GPUs,
+# since the lattice generation runs in about real-time, so takes of the order of
+# 1000 hours of CPU time.
+# 
+. cmd.sh
```
Comment (vijayaditya) [inline egs/librispeech/s5/local/chain/run_tdnn_6z_discriminative.sh]: delete line.

File: egs/librispeech/s5/local/nnet3/run_ivector_common.sh [hunk 2]
```diff
@@ -0,0 +1,142 @@
+#!/bin/bash
+
+# this script contains some common (shared) parts of the run_nnet*.sh scripts.
+
+. cmd.sh
```
Comment (vijayaditya) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: delete line

File: egs/librispeech/s5/local/nnet3/run_ivector_common.sh [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  # this shows how you can split across multiple file-systems.  we'll split the
+  # MFCC dir across multiple locations.  You might want to be careful here, if you
+  # have multiple copies of Kaldi checked out and run the same recipe, not to let
+  # them overwrite each other.
+  mfccdir=mfcc_hires
+  if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d $mfccdir/storage ]; then
+    utils/create_split_dir.pl /export/b0{1,2,3,4}/$USER/kaldi-data/egs/librispeech-$(date +'%m_%d_%H_%M')/s5/$mfccdir/storage $mfccdir/storage
+  fi
+
+  for datadir in $train_set test_clean test_other dev_clean dev_other; do
```
Comment (vijayaditya) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: do volume perturbation. Use egs/wsj/s5/utils/data/perturb_data_dir_volume.sh

Comment (vijayaditya) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: You need not rerun the experiments.

**Comment by vijayaditya** [comment]
make the suggested changes, I will merge after that.

File: egs/librispeech/s5/local/nnet3/run_ivector_common.sh (hunk 3 above)
Comment (freewym) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: Done. I assume volume perturbation is applied only on training data.

Comment (vijayaditya) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: Yes its just done on the training data. If you address my other comments I will merge this.

Comment (freewym) [inline egs/librispeech/s5/local/nnet3/run_ivector_common.sh]: Done

**Comment by vijayaditya** [comment]
You did not add the results.

**Comment by freewym** [comment]
Updated the RESULT file.

File: egs/librispeech/s5/local/chain/run_chain_common.sh [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+# output directory names
+dir=
+treedir=
+lang=
+min_seg_len=
+# End configuration section.
+echo "$0 $@"  # Print the command line for logging
+
+. cmd.sh
```
Comment (vijayaditya) [inline egs/librispeech/s5/local/chain/run_chain_common.sh]: you should source cmd.sh in the current directory. The cmd.sh file might not be available in the path.

so `. cmd.sh` should be changed to `. ./cmd.sh`.

This change was pointed out yesterday, but you might have missed the comment. You could actually monitor the conversation in github to find the latest changes.

Comment (freewym) [inline egs/librispeech/s5/local/chain/run_chain_common.sh]: Sorry. Fixed.

**Comment by vijayaditya** [comment]
Only minor changes are required and sMBR results need to added. These can be added later. I am merging now.

(For more detailed review comments see #645)

**Comment by freewym** [comment]
Update: Further increasing the relu-dim in tdnn+chain model from 725 to 896 improve the WER on "clean" data a bit, but WER on "other" data get a little worse. So overall we might just keep the current setting where relu-dim is 725? 

relu-dim=896, num_params=23.6M (18.3M after excluding the xent branch)
%WER 3.82 [ 2077 / 54402, 282 ins, 194 del, 1601 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_clean_fglarge/wer_11_0.5
%WER 3.96 [ 2156 / 54402, 280 ins, 217 del, 1659 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_clean_tglarge/wer_12_0.5
%WER 5.05 [ 2746 / 54402, 335 ins, 270 del, 2141 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_clean_tgmed/wer_12_0.0
%WER 5.69 [ 3093 / 54402, 328 ins, 353 del, 2412 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_clean_tgsmall/wer_12_0.0
%WER 10.53 [ 5367 / 50948, 668 ins, 589 del, 4110 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_other_fglarge/wer_15_0.0
%WER 11.06 [ 5633 / 50948, 700 ins, 644 del, 4289 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_other_tglarge/wer_15_0.0
%WER 13.40 [ 6829 / 50948, 719 ins, 915 del, 5195 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_other_tgmed/wer_14_0.0
%WER 14.53 [ 7403 / 50948, 653 ins, 1139 del, 5611 sub ] exp/chain/tdnn_6z_relu896_sp/decode_dev_other_tgsmall/wer_15_0.0
%WER 4.24 [ 2230 / 52576, 297 ins, 222 del, 1711 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_clean_fglarge/wer_11_1.0
%WER 4.39 [ 2310 / 52576, 307 ins, 229 del, 1774 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_clean_tglarge/wer_12_0.5
%WER 5.41 [ 2842 / 52576, 385 ins, 266 del, 2191 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_clean_tgmed/wer_12_0.0
%WER 6.02 [ 3166 / 52576, 374 ins, 347 del, 2445 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_clean_tgsmall/wer_12_0.0
%WER 10.96 [ 5737 / 52343, 660 ins, 699 del, 4378 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_other_fglarge/wer_13_0.5
%WER 11.31 [ 5922 / 52343, 765 ins, 644 del, 4513 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_other_tglarge/wer_14_0.0
%WER 13.56 [ 7099 / 52343, 787 ins, 963 del, 5349 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_other_tgmed/wer_14_0.0
%WER 14.95 [ 7824 / 52343, 760 ins, 1191 del, 5873 sub ] exp/chain/tdnn_6z_relu896_sp/decode_test_other_tgsmall/wer_14_0.0

-------------------------------------------------------------------------------------------

relu-dim=725, num_params=16.8M (12.7M after excluding the xent branch)
%WER 3.92 [ 2131 / 54402, 290 ins, 197 del, 1644 sub ] exp/chain/tdnn_6z_sp/decode_dev_clean_fglarge/wer_11_0.5
%WER 4.09 [ 2227 / 54402, 337 ins, 176 del, 1714 sub ] exp/chain/tdnn_6z_sp/decode_dev_clean_tglarge/wer_11_0.0
%WER 5.11 [ 2781 / 54402, 329 ins, 300 del, 2152 sub ] exp/chain/tdnn_6z_sp/decode_dev_clean_tgmed/wer_12_0.0
%WER 5.83 [ 3172 / 54402, 335 ins, 372 del, 2465 sub ] exp/chain/tdnn_6z_sp/decode_dev_clean_tgsmall/wer_12_0.0
%WER 10.43 [ 5314 / 50948, 528 ins, 697 del, 4089 sub ] exp/chain/tdnn_6z_sp/decode_dev_other_fglarge/wer_14_0.5
%WER 10.95 [ 5581 / 50948, 546 ins, 764 del, 4271 sub ] exp/chain/tdnn_6z_sp/decode_dev_other_tglarge/wer_14_0.5
%WER 13.20 [ 6723 / 50948, 676 ins, 858 del, 5189 sub ] exp/chain/tdnn_6z_sp/decode_dev_other_tgmed/wer_13_0.0
%WER 14.56 [ 7419 / 50948, 715 ins, 1003 del, 5701 sub ] exp/chain/tdnn_6z_sp/decode_dev_other_tgsmall/wer_13_0.0
%WER 4.28 [ 2251 / 52576, 292 ins, 238 del, 1721 sub ] exp/chain/tdnn_6z_sp/decode_test_clean_fglarge/wer_11_1.0
%WER 4.47 [ 2349 / 52576, 342 ins, 225 del, 1782 sub ] exp/chain/tdnn_6z_sp/decode_test_clean_tglarge/wer_11_0.5
%WER 5.55 [ 2917 / 52576, 366 ins, 314 del, 2237 sub ] exp/chain/tdnn_6z_sp/decode_test_clean_tgmed/wer_13_0.0
%WER 6.20 [ 3259 / 52576, 383 ins, 381 del, 2495 sub ] exp/chain/tdnn_6z_sp/decode_test_clean_tgsmall/wer_12_0.0
%WER 10.76 [ 5634 / 52343, 643 ins, 672 del, 4319 sub ] exp/chain/tdnn_6z_sp/decode_test_other_fglarge/wer_12_0.5
%WER 11.20 [ 5864 / 52343, 619 ins, 781 del, 4464 sub ] exp/chain/tdnn_6z_sp/decode_test_other_tglarge/wer_13_0.5
%WER 13.47 [ 7051 / 52343, 733 ins, 933 del, 5385 sub ] exp/chain/tdnn_6z_sp/decode_test_other_tgmed/wer_13_0.0
%WER 14.73 [ 7710 / 52343, 662 ins, 1209 del, 5839 sub ] exp/chain/tdnn_6z_sp/decode_test_other_tgsmall/wer_14_0.0

**Comment by danpovey** [comment]
Thanks.  Yes, just keep the current setting then.

**Pull request merged**